go/tools/asthelpergen/asthelpergen.go

```diff
@@ -247,13 +306,137 @@ func GenerateASTHelpers(packagePatterns []string, rootIface, exceptCloneType str
 		return types.Implements(t, iface)
 	}
 	rewriter := newRewriterGen(interestingType, nt.Obj().Name())
-	clone := newCloneGen(iface, scope, exceptCloneType)
+	generator := newGenerator(loaded[0].Module, loaded[0].TypesSizes, nt, rewriter)
+	generator.gens2 = append(generator.gens2, &equalsGen{})
+	generator.gens2 = append(generator.gens2, newCloneGen(exceptCloneType))
+	generator.gens2 = append(generator.gens2, &visitGen{})
 
-	generator := newGenerator(loaded[0].Module, loaded[0].TypesSizes, nt, rewriter, clone)
 	it, err := generator.GenerateCode()
 	if err != nil {
 		return nil, err
 	}
 
 	return it, nil
 }
+
+var _ generatorSPI = (*astHelperGen)(nil)
+
+func (gen *astHelperGen) scope() *types.Scope {
+	return gen._scope
+}
+
+func (gen *astHelperGen) addType(t types.Type) {
+	gen.todo = append(gen.todo, t)
+}
+
+type methodType int
+
+const (
+	clone methodType = iota
+	equals
+	visit
+)
+
+func (gen *astHelperGen) addFunc(name string, typ methodType, code jen.Code) {
+	var comment string
+	switch typ {
+	case clone:
+		comment = " creates a deep clone of the input."
+	case equals:
+		comment = " does deep equals between the two objects."
+	case visit:
+		comment = " will visit all parts of the AST"
+	}
+	gen.methods = append(gen.methods, jen.Comment(name+comment), code)
+}
+
+func (gen *astHelperGen) createFile(pkgName string) (string, *jen.File) {
+	out := jen.NewFile(pkgName)
+	out.HeaderComment(licenseFileHeader)
+	out.HeaderComment("Code generated by ASTHelperGen. DO NOT EDIT.")
+	alreadyDone := map[string]bool{}
+	for len(gen.todo) > 0 {
+		t := gen.todo[0]
+		underlying := t.Underlying()
+		typeName := printableTypeName(t)
+		gen.todo = gen.todo[1:]
+
+		if alreadyDone[typeName] {
+			continue
+		}
+
+		switch underlying := underlying.(type) {
+		case *types.Interface:
+			gen.allGenerators(func(g generator2) error {
+				return g.interfaceMethod(t, underlying, gen)
+			})
+		case *types.Slice:
+			gen.allGenerators(func(g generator2) error {
+				return g.sliceMethod(t, underlying, gen)
+			})
+		case *types.Struct:
+			gen.allGenerators(func(g generator2) error {
+				return g.structMethod(t, underlying, gen)
+			})
+		case *types.Pointer:
+			ptrToType := underlying.Elem().Underlying()
+
+			switch ptrToType := ptrToType.(type) {
+			case *types.Struct:
+				gen.allGenerators(func(g generator2) error {
+					return g.ptrToStructMethod(t, ptrToType, gen)
+				})
+			case *types.Basic:
+				gen.allGenerators(func(g generator2) error {
+					return g.ptrToBasicMethod(t, ptrToType, gen)
+				})
+			default:
+				panic(fmt.Sprintf("%T", ptrToType))
```

Comment:
nit: we can do log.fatal here, or is this intentional?

Reply:
+1. If there's a good reason to panic, please document it.